### PR TITLE
implement `__deepcopy__`

### PR DESCRIPTION
**Pull request by kornerc**
**Fixes issue**: #1185 

### Changes proposed in this PR include:
- Implement `__deepcopy__` for `PowerGridModel`
- Write unit test for this

### CODE REVIEW

File: src/power_grid_model/_core/power_grid_model.py [hunk 1]
```diff
@@ -108,9 +108,18 @@ def copy(self) -> "PowerGridModel":
         new_model._all_component_count = self._all_component_count
         return new_model
 
-    def __copy__(self):
+    def __copy__(self) -> "PowerGridModel":
         return self.copy()
 
+    def __deepcopy__(self, memo: dict[int, Any]) -> "PowerGridModel":
+        # PowerGridModel.copy makes already a deepcopy
```
Comment (TonyXiang8787) [inline src/power_grid_model/_core/power_grid_model.py:115]: Do you need to check if `id(self)` is already in the `memo`? Otherwise we do a lot of copies if this object is referenced multiple times in upper layer object?

File: tests/unit/test_power_grid_model.py [hunk 2]
```diff
@@ -176,6 +176,12 @@ def test_copy_model(model: PowerGridModel, sym_output):
     compare_result(result, sym_output, rtol=0.0, atol=1e-8)
 
 
+def test_deepcopy_model(model: PowerGridModel, sym_output):
+    model_2 = deepcopy(model)
+    result = model_2.calculate_power_flow()
+    compare_result(result, sym_output, rtol=0.0, atol=1e-8)
```
Comment (TonyXiang8787) [inline tests/unit/test_power_grid_model.py:207]: Please try to test a `deepcopy` of a `list` with two references to the same model to make sure that the `memo` works as intended.

```python
model_list = [model, model]
new_model_list = deepcopy(model_list)
assert id(new_model_list[0]) == id(new_model_list[1])
assert id(new_model_list[0]) != id(model)
```

Comment (kornerc) [reply tests/unit/test_power_grid_model.py:207]: I've changed the test to cover more edge cases.
Please let me know if this is okay.

File: src/power_grid_model/_core/power_grid_model.py (hunk 1 above)
Comment (kornerc) [reply src/power_grid_model/_core/power_grid_model.py:115]: This is not needed since deepcopy checks [this already](https://github.com/python/cpython/blob/6f988b08d122e44848e89c04ad1e10c25d072cc7/Lib/copy.py#L126).
In our case it would even be possible to avoid setting memo entirely without any negative impact since deeopcopy is [taking care about that](https://github.com/python/cpython/blob/6f988b08d122e44848e89c04ad1e10c25d072cc7/Lib/copy.py#L161) too.
However, to the best of my knowledge it is recommended to set memo in custom `__deepcopy__` implementations.

**Review by TonyXiang8787 — APPROVED** [review]
@kornerc Thanks for your contribution. After CI check passes, the PR will be merged.
